x/ecocredit/server/basket/put.go

```diff
@@ -142,36 +139,38 @@ func (k Keeper) canBasketAcceptCredit(ctx context.Context, basket *api.Basket, b
 	}
 
 	// check credit type match
-	requiredCreditType := basket.CreditTypeAbbrev
-	res, err := k.ecocreditKeeper.ClassInfo(ctx, &ecocredit.QueryClassInfoRequest{ClassId: classId})
+	class, err := k.coreStore.ClassInfoTable().GetByName(ctx, classId)
 	if err != nil {
 		return err
 	}
-	gotCreditType := res.Info.CreditType.Abbreviation
-	if requiredCreditType != gotCreditType {
-		return errInvalidReq.Wrapf("cannot use credit of type %s in a basket that requires credit type %s", gotCreditType, requiredCreditType)
+	if class.CreditType != basket.CreditTypeAbbrev {
+		return errInvalidReq.Wrapf("basket requires credit type %s but a credit with type %s was given", basket.CreditTypeAbbrev, class.CreditType)
 	}
+
 	return nil
 }
 
-// transferToBasket updates the balance of the user in the legacy KVStore as well as the basket's balance in the ORM.
-func (k Keeper) transferToBasket(ctx context.Context, sender sdk.AccAddress, amt regenmath.Dec, basket *api.Basket, batchInfo *ecocredit.BatchInfo) error {
-	sdkCtx := sdk.UnwrapSDKContext(ctx)
-	store := sdkCtx.KVStore(k.storeKey)
-
-	// update the user balance
-	userBalanceKey := ecocredit.TradableBalanceKey(sender, ecocredit.BatchDenomT(batchInfo.BatchDenom))
-	userBalance, err := ecocredit.GetDecimal(store, userBalanceKey)
+// transferToBasket moves credits from the user's tradable balance, into the basket's balance
+func (k Keeper) transferToBasket(ctx context.Context, sender sdk.AccAddress, amt regenmath.Dec, basket *api.Basket, batchInfo *ecoApi.BatchInfo) error {
+	// update user balance, subtracting from their tradable balance
+	userBal, err := k.coreStore.BatchBalanceTable().Get(ctx, sender, batchInfo.Id)
+	if err != nil {
+		return ecocredit.ErrInsufficientFunds.Wrapf("could not get batch %s balance for %s", batchInfo.BatchDenom, sender.String())
+	}
+	tradable, err := regenmath.NewPositiveDecFromString(userBal.Tradable)
 	if err != nil {
 		return err
 	}
-	newUserBalance, err := regenmath.SafeSubBalance(userBalance, amt)
+	newTradable, err := regenmath.SafeSubBalance(tradable, amt)
 	if err != nil {
+		return ecocredit.ErrInsufficientFunds.Wrapf("cannot put %v credits into the basket with a balance of %v: %s", amt, tradable, err.Error())
+	}
+	userBal.Tradable = newTradable.String()
+	if err = k.coreStore.BatchBalanceTable().Save(ctx, userBal); err != nil {
```

Comment:
Should this be `Update` rather than `Save` ? A user will already have a batch balance.

Reply:
https://github.com/regen-network/regen-ledger/pull/935/commits/cde0f1916f46e2cfac9535cb8629abd138ad2f52